### PR TITLE
dumpling: make prom registry configurable and use const labels

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/34948

Problem Summary:

https://github.com/pingcap/tidb/pull/34946 introduced `promutil.Factory` to support wrapping custom logic when creating metrics. However, the caller doesn't want to register metrics in the dumpling. This leads us to make particular judgments for metric registration, which brings much complexity to the dumpling.

### What is changed and how it works?

* Add `promutil.Registry`  and provide a `noopRegistry` implementation.
* Use const labels during defining metrics.
* Unify metrics registration for both standalone and library usage.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: dumpling/export/dump.go [hunk 1]
```diff
@@ -81,11 +81,8 @@ func NewDumper(ctx context.Context, conf *Config) (*Dumper, error) {
 		selectTiDBTableRegionFunc: selectTiDBTableRegion,
 	}
 
-	if conf.PromFactory == nil {
-		d.metrics = defaultMetrics
-	} else {
-		d.metrics = newMetrics(conf.PromFactory, []string{})
-	}
+	d.metrics = newMetrics(conf.PromFactory, conf.Labels)
+	d.metrics.registerTo(conf.PromRegistry)
```
Comment (lance6716) [inline dumpling/export/dump.go:85]: I think we don't need this call and `conf.PromRegistry`. 
For Factory injected by dataflow engine, the Factory will automatically register the metrics variables when NewXXX in `newMetrics`.  So for other cases we can also imitate this behaviour. 
- For standalone dumpling binary, we let the default Factory become something like NewAutoRegisterFactory(NewDefaultFactory, ...)
where NewAutoRegisterFactory is at https://github.com/pingcap/tiflow/blob/329230aa9359836577f6a638168b4e717c045c24/engine/pkg/promutil/registry.go#L99

- And for DM's case DM should let the Factory argument has NewAutoRegisterFactory wrapped

Comment (lance6716) [reply dumpling/export/dump.go:85]: AutoRegisterFactory need a explicit Registry, that Registry can call `Unregister` when dumpling is about to close. https://github.com/pingcap/tiflow/blob/329230aa9359836577f6a638168b4e717c045c24/engine/pkg/promutil/registry.go#L79 In fact, offical prometheus also has a "AutoRegisterFactory" https://pkg.go.dev/github.com/prometheus/client_golang/prometheus/promauto , but compared to it, ours can handle Unregister.

in my understanding, the decision is more about other applications which want to use dumpling/lightning as a library? I now a bit agree that explitly requiring two things is less error-prone. Will review later.

also cc @maxshuang

File: dumpling/export/writer_serial_test.go [hunk 2]
```diff
@@ -52,7 +52,8 @@ func TestWriteInsert(t *testing.T) {
 	bf := storage.NewBufferWriter()
 
 	conf := configForWriteSQL(cfg, UnspecifiedSize, UnspecifiedSize)
-	n, err := WriteInsert(tcontext.Background(), conf, tableIR, tableIR, bf, defaultMetrics)
+	m := newMetrics(conf.PromFactory, conf.Labels)
```
Comment (lance6716) [inline dumpling/export/writer_serial_test.go:55]: seems we can remove cfg.Labels now because DM should use constLabels Factory, or we can do it later

**Review by lance6716 — APPROVED** [review]

File: dumpling/export/writer_serial_test.go (hunk 2 above)
Comment (gozssky) [reply dumpling/export/writer_serial_test.go:55]: It's useful for someone who just wants to use the original prometheus factory.

**Pull request merged**